File: labs/typescript/tests/lab.ts
// Shared plumbing for the lab tests. Students should not need to edit this file.

export const BASE_URL = process.env.MOCKHUB_URL ?? "https://mockhub.kousenit.com";
export const ACP_KEY = process.env.MOCKHUB_ACP_KEY ?? "mockhub-dev-key";
export const EMAIL = process.env.MOCKHUB_EMAIL ?? "buyer@mockhub.com";
export const PASSWORD = process.env.MOCKHUB_PASSWORD ?? "buyer123";

// the AGENT's credential: an API key, no user login
export const AGENT_HEADERS = { "X-API-Key": ACP_KEY };

async function http(path: string, init?: RequestInit): Promise<Response> {
  try {
    return await fetch(BASE_URL + path, init);
  } catch (e) {
    throw new Error(
      `Could not reach MockHub at ${BASE_URL} — check your network, then ask in chat. (${e})`,
    );
  }
}

export const get = (path: string, headers: Record<string, string> = {}) =>
  http(path, { headers });

export const post = (path: string, body: unknown, headers: Record<string, string> = {}) =>
  http(path, {
    method: "POST",
    headers: { "Content-Type": "application/json", ...headers },
    body: JSON.stringify(body),
  });

/** The HUMAN's credential: a JWT from the buyer's username and password. */
export async function login(): Promise<string> {
  const response = await post("/api/v1/auth/login", { email: EMAIL, password: PASSWORD });
  if (!response.ok) throw new Error(`login failed: ${response.status}`);
  return (await response.json()).accessToken;
}

/** Unique per run, so students sharing the demo account never collide. */
export const freshAgentId = () =>
  `lab-agent-${crypto.randomUUID().replaceAll("-", "").slice(0, 12)}`;

export async function createMandate(
  token: string,
  agentId: string,
  maxSpendPerTransaction: string,
  approvalMode: "AUTO_PURCHASE" | "APPROVAL_REQUIRED",
) {
  const response = await post(
    "/api/v1/my/mandates",
    { agentId, scope: "PURCHASE", maxSpendPerTransaction, approvalMode },
    { Authorization: `Bearer ${token}` },
  );
  if (!response.ok) throw new Error(`createMandate failed: ${response.status}`);
  return response.json();
}

export const revokeMandate = (token: string, mandateId: string) =>
  http(`/api/v1/my/mandates/${mandateId}`, {
    method: "DELETE",
    headers: { Authorization: `Bearer ${token}` },
  });

/** Any listing costing more than a dollar; a random page avoids seat contention. */
export async function pickAListing() {
  const page = Math.floor(Math.random() * 5);
  const response = await get(
    `/acp/v1/listings?size=10&minPrice=10&page=${page}`,
    AGENT_HEADERS,
  );
  if (!response.ok) throw new Error(`listings failed: ${response.status}`);
  const listings = (await response.json()).content;
  if (!listings?.length) throw new Error("MockHub should have listings for sale");
  return listings[Math.floor(Math.random() * listings.length)];
}

/** The agent's purchase path: ACP checkout, carrying its mandate. */
export const agentTriesToBuy = (agentId: string, mandateId: string, listingId: number) =>
  post(
    "/acp/v1/checkout",
    {
      buyerEmail: EMAIL,
      lineItems: [{ listingId, quantity: 1 }],
      agentId,
      mandateId,
    },
    AGENT_HEADERS,
  );

export const agentTriesToComplete = (checkoutId: string, agentId: string, mandateId: string) =>
  post(
    `/acp/v1/checkout/${checkoutId}/complete`,
    { agentId, mandateId },
    { ...AGENT_HEADERS, "X-Buyer-Email": EMAIL },
  );

export const cancelCheckout = (checkoutId: string, agentId: string, mandateId: string) =>
  post(
    `/acp/v1/checkout/${checkoutId}/cancel`,
    { agentId, mandateId },
    { ...AGENT_HEADERS, "X-Buyer-Email": EMAIL },
  );

File: labs/python/lab.py
"""Shared plumbing for the lab tests. Students should not need to edit this file."""

import os
import uuid

import requests

BASE_URL = os.environ.get("MOCKHUB_URL", "https://mockhub.kousenit.com")
ACP_KEY = os.environ.get("MOCKHUB_ACP_KEY", "mockhub-dev-key")
EMAIL = os.environ.get("MOCKHUB_EMAIL", "buyer@mockhub.com")
PASSWORD = os.environ.get("MOCKHUB_PASSWORD", "buyer123")

AGENT_HEADERS = {"X-API-Key": ACP_KEY}  # the AGENT's credential: an API key, no user login


def login() -> str:
    """The HUMAN's credential: a JWT from the buyer's username and password."""
    response = requests.post(
        f"{BASE_URL}/api/v1/auth/login",
        json={"email": EMAIL, "password": PASSWORD},
        timeout=15,
    )
    response.raise_for_status()
    return response.json()["accessToken"]


def fresh_agent_id() -> str:
    """Unique per run, so students sharing the demo account never collide."""
    return f"lab-agent-{uuid.uuid4().hex[:12]}"


def create_mandate(token: str, agent_id: str, **fields) -> dict:
    body = {"agentId": agent_id, "scope": "PURCHASE", **fields}
    response = requests.post(
        f"{BASE_URL}/api/v1/my/mandates",
        json=body,
        headers={"Authorization": f"Bearer {token}"},
        timeout=15,
    )
    response.raise_for_status()
    return response.json()

def revoke_mandate(token: str, mandate_id: str) -> None:
    requests.delete(
        f"{BASE_URL}/api/v1/my/mandates/{mandate_id}",
        headers={"Authorization": f"Bearer {token}"},
        timeout=15,
    )


def pick_a_listing() -> dict:
    """Any listing costing more than a dollar; a random page avoids seat contention."""
    page = uuid.uuid4().int % 5
    response = requests.get(
        f"{BASE_URL}/acp/v1/listings",
        params={"size": 10, "page": page, "minPrice": 10},
        headers=AGENT_HEADERS,
        timeout=15,
    )
    response.raise_for_status()
    listings = response.json()["content"]
    assert listings, "MockHub should have listings for sale"
    return listings[uuid.uuid4().int % len(listings)]


def agent_tries_to_buy(agent_id: str, mandate_id: str, listing_id: int) -> requests.Response:
    """The agent's purchase path: ACP checkout, carrying its mandate."""
    return requests.post(
        f"{BASE_URL}/acp/v1/checkout",
        json={
            "buyerEmail": EMAIL,
            "lineItems": [{"listingId": listing_id, "quantity": 1}],
            "agentId": agent_id,
            "mandateId": mandate_id,
        },
        headers=AGENT_HEADERS,
        timeout=15,
    )


def agent_tries_to_complete(checkout_id: str, agent_id: str, mandate_id: str,
                            approval_id: str | None = None) -> requests.Response:
    body = {"agentId": agent_id, "mandateId": mandate_id}
    if approval_id:
        body["approvalId"] = approval_id
    return requests.post(
        f"{BASE_URL}/acp/v1/checkout/{checkout_id}/complete",
        json=body,
        headers={**AGENT_HEADERS, "X-Buyer-Email": EMAIL},
        timeout=15,
    )


def cancel_checkout(checkout_id: str) -> None:
    requests.post(
        f"{BASE_URL}/acp/v1/checkout/{checkout_id}/cancel",
        headers={**AGENT_HEADERS, "X-Buyer-Email": EMAIL},
        timeout=15,
    )
